### ISSUE TITLE
[FR] Add Pandas category dtype to mlflow.types.schema

### ISSUE THREAD

**Issue by henriqueluzz** [mlflow/mlflow#3849]
## Willingness to contribute
The MLflow Community encourages new feature contributions. Would you or another member of your organization be willing to contribute an implementation of this feature (either as an MLflow Plugin or an enhancement to the MLflow code base)?

- [ ] Yes. I can contribute this feature independently.
- [X] Yes. I would be willing to contribute this feature with guidance from the MLflow community.
- [ ] No. I cannot contribute this feature at this time.

## Proposal Summary
Add Pandas category dtype, to mlflow.types.schema().

I currently have a ML model based on LightGBM running on production environment that contains categorical variables. Instead of label-encoding or OHE the categorical variables, I'm using Pandas dataframe with categorical columns set to be of the categorical dtype.

` df['some_categorical_column'] = df['some_categorical_column'].astype('category') .`

I'm trying to deploy this model on Databricks Serve Model that uses MLFlow as backend, but currently this category dtype is not supported for model signature according to the available dtypes on [https://www.mlflow.org/docs/latest/_modules/mlflow/types/schema.html](here), so I'm unable to correctly cast the prediction Dataframe accordingly to the training phase.

It makes me think that deploying models that were trained with LightGBM using Pandas category dtypes, cannot be correctly deployed since this dtype is not availabe, so it would occur some casting error during the prediction leading to incorrect scores/predictions. 

I could of course OHE the data and outline the problem, but I'm trying to deploy the current model into MLFlow.
I'm not sure if this is a FR or something else.

![bug2](https://user-images.githubusercontent.com/3441278/102220152-2f68b900-3ebf-11eb-9417-871fd027cca8.PNG)

### What component(s), interfaces, languages, and integrations does this feature affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [X] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

**Comment by dmatrix** [comment]
@tomasatdatabricks is this something that should be supported?

**Comment by tomasatdatabricks** [comment]
Possibly yes, but it needs a bit of work. 

The tricky part about categoricals is that the dictionary can be different between training and inference data. So you need to store the mapping seen during training and make sure you map the data the same way at inference time. You also need to handle categories unseen during training. We would also need to figure out where to store these additional dictionaries - probably as additional artifacts. It can be done, but it is not clear how often would people use this instead of using something like scikit pipeline.

**Comment by olbapjose** [comment]
Any advances on this? I think it would be useful as there are some algorithms which support null values directly (e.g. XGBoost, which also supports categorical variables in a straightforward way from now on).

**Comment by frsnjung** [comment]
I would also like to have this feature to be able to deploy LGBM models easily that use features declared as categorical to use optimal bin sizes https://lightgbm.readthedocs.io/en/latest/Advanced-Topics.html#categorical-feature-support ).

**Comment by eliwoods** [comment]
I know this is an old issue, but would love to see support for this as well. `auto-sklearn` makes use of the category dtype as well to add the list of libraries that directly leverage this functionality.

As an intermediate solution, it could be possible to accept the category type by casting to a generic `object` dtype and ignoring the dictionaries of category values. This obviously isn't ideal, but this is the local workaround that I am currently doing to successfully pass my inputs to `infer_signature`.

**Comment by dbczumar** [comment]
@tomasatdatabricks What are your latest thoughts here?

**Comment by tmg-ling** [comment]
Any updates for this request?

**Comment by atolnix** [comment]
Any updates on this feature?

**Comment by lene-hansen** [comment]
Also interested in an update on this :)

**Comment by RashidBakirov** [comment]
Would love to have an update on this as well, spent quite a bit of time trying to debug an issue with serving LightGBM model with categoricals via MLFlow and finding out that this apparently is not possible.

**Comment by steffyd** [comment]
Would this be related to an issue that I have with running evaluate on an XGBoost model? Getting the error:

File "C:\Python310\lib\site-packages\mlflow\models\utils.py", line 577, in _enforce_named_col_schema
    new_pf_input[x] = _enforce_mlflow_datatype(x, pf_input[x], input_types[x])
  File "C:\Python310\lib\site-packages\mlflow\models\utils.py", line 554, in _enforce_mlflow_datatype
    raise MlflowException(
mlflow.exceptions.MlflowException: Incompatible input types for column Feature_RecruiterState. Can not safely convert category to <U0.

**Comment by RashidBakirov** [comment]
As an update to this, using "category" type works with MLFlow. 

> Would this be related to an issue that I have with running evaluate on an XGBoost model? Getting the error:
> 
> File "C:\Python310\lib\site-packages\mlflow\models\utils.py", line 577, in _enforce_named_col_schema new_pf_input[x] = _enforce_mlflow_datatype(x, pf_input[x], input_types[x]) File "C:\Python310\lib\site-packages\mlflow\models\utils.py", line 554, in _enforce_mlflow_datatype raise MlflowException( mlflow.exceptions.MlflowException: Incompatible input types for column Feature_RecruiterState. Can not safely convert category to <U0.

I had the same error with LightGBM. In the end I managed to fix this specifying categorical features explicitly, using  "categorical_feature" argument.

**Comment by rtambone** [comment]
> As an update to this, using "category" type works with MLFlow.
> 
> > Would this be related to an issue that I have with running evaluate on an XGBoost model? Getting the error:
> > File "C:\Python310\lib\site-packages\mlflow\models\utils.py", line 577, in _enforce_named_col_schema new_pf_input[x] = _enforce_mlflow_datatype(x, pf_input[x], input_types[x]) File "C:\Python310\lib\site-packages\mlflow\models\utils.py", line 554, in _enforce_mlflow_datatype raise MlflowException( mlflow.exceptions.MlflowException: Incompatible input types for column Feature_RecruiterState. Can not safely convert category to <U0.
> 
> I had the same error with LightGBM. In the end I managed to fix this specifying categorical features explicitly, using "categorical_feature" argument.

Sorry, I also use the categorical features explictly but I got the same error. 
Basically I fit the LGB model on a pandas Dataframe with categorical columns. Then I log the model in the Databricks UC via MLflow. After that I would like to use MLflow APIs to perform model evaluation/validation. At this time when the model is loaded via mlflow as pyfunc, it tries to enforce schema and I got the error

**Comment by kylebeni** [comment]
It would be great if this can be implemented. Categorical features are ubiquitous with lightgbm and their implementation is more powerful than other encoding methods, so MLFlow's support for lightgbm (and other libraries) is incomplete without support for categoricals.

**Comment by kylebeni** [comment]
UPDATE:

This is possible with a workaround:

During **training**, set the input schema for any categoricals to `mlflow.types.DataType.string`. 
⚠️ **Important:** This maps to a pandas `object` type. So your input data for the categorical field must be an `object` dtype during prediction. This ensures that it passes the schema check.

Then, using a custom pyfunc model class, set up a preprocessing step to convert the column to a categorical from an object. This ensures that the data passed to lightgbm is of the correct pandas `category` type.

**Important:** make sure to set `categorical_feature` arg is passed during `fit()`.

## Fit Model
```python
    model.fit(
        X=X_train,
        y=y_train,
        eval_set=[(X_train, y_train), (X_valid, y_valid)],
        callbacks=[e_stop_callback],
        categorical_feature=["cat_col"],
    )

# Save LGBM
model_directory = "/tmp/lgbm_model.joblib"
dump(model, model_directory)
```

## Define schema manually
```python
mlf_input_schema = Schema(
    [
        ColSpec("string", "cat_col", required=False),
    ]
)
```

## Log Custom Model
```python
class ModelWrapper(PythonModel):
        def __init__(self):
            self.model = None

        def load_context(self, context):
            from joblib import load

            self.model = load(context.artifacts["model_path"])

        def preprocess_inputs(self, inputs):
            return inputs.astype({"cat_col": "category"})

        def predict(self, context, model_input, params=None):
            params = params or {"predict_method": "predict_proba"}
            predict_method = params.get("predict_method")

            model_input = self.preprocess_inputs(inputs=model_input)

            if predict_method == "predict":
                return self.model.predict(model_input)
            elif predict_method == "predict_proba":
                return self.model.predict_proba(model_input)[:, 1]
            else:
                raise ValueError(
                    f"The prediction method '{predict_method}' is not supported."
                )

artifacts = {"model_path": model_directory}

input_schema = mlf_input_schema

param_schema = ...

signature = ModelSignature(inputs=input_schema, params=param_schema)

log_model(
    artifact_path="model",
    python_model=ModelWrapper(),
    input_example=X_sample,
    signature=signature,
    artifacts=artifacts,
    )
```

**Comment by tuomijal** [comment]
Thank you @kylebeni. I am also interested in this.

**Comment by tdhooghe** [comment]
Same! It also seems that when logging the model, mlflow just forgets the `enable_categorical` setting and just sets it too false..
